# **1️⃣ Eigenvalues & Eigenvectors: Definition, Linear Transformations & Applications 🔢📊**  

## **💡 Real-Life Analogy: Stretching and Rotating an Object 🎭**  

Imagine you are **stretching or rotating a rubber sheet**:  
- Some **directions stay unchanged** (they just stretch or shrink) → These are **eigenvectors**.  
- The **amount they stretch or shrink** is the **eigenvalue**.  
- The rest of the rubber may distort, but **eigenvectors remain on their original lines**, just scaled.  

📌 **Eigenvalues and eigenvectors help us understand how transformations affect spaces!**  

## **📌 What are Eigenvalues and Eigenvectors?**  

✅ **Eigenvectors** are **special vectors** that, when a linear transformation (matrix) is applied, **do not change direction**, only their **magnitude changes**.  
✅ **Eigenvalues** are the **scaling factors** that describe how much the eigenvector is **stretched or shrunk** during the transformation.  

## **📌 Mathematical Definition:**  

For a **square matrix** $A$, an **eigenvector** $v$ and **eigenvalue** $\lambda$ satisfy:  
  
$$  
A v = \lambda v  
$$  
  
Where:  
- $A$ = **Transformation matrix**.  
- $v$ = **Eigenvector (a nonzero vector that remains in the same direction)**.  
- $\lambda$ = **Eigenvalue (scaling factor of the eigenvector)**.  

## **📊 Example 1: Simple 2D Linear Transformation**  

📌 **Given Transformation Matrix:**  
$$  
A = \begin{bmatrix} 2 & 1 \\ 1 & 2 \end{bmatrix}  
$$  

📌 **Finding Eigenvalues:**  
Solve $\det(A - \lambda I) = 0$, where $I$ is the identity matrix.  
  
$$  
\begin{vmatrix} 2 - \lambda & 1 \\ 1 & 2 - \lambda \end{vmatrix} = 0  
$$  

Expanding the determinant:  
  
$$  
(2-\lambda)(2-\lambda) - (1)(1) = 0  
$$  

$$  
(4 - 4\lambda + \lambda^2) - 1 = 0  
$$  

$$  
\lambda^2 - 4\lambda + 3 = 0  
$$  

$$  
(\lambda - 3)(\lambda - 1) = 0  
$$  

✅ **Eigenvalues:** $\lambda_1 = 3$, $\lambda_2 = 1$.  

📌 **Finding Eigenvectors:** Solve $(A - \lambda I)v = 0$ for each $\lambda$.  

For $\lambda_1 = 3$:  
$$  
\begin{bmatrix} -1 & 1 \\ 1 & -1 \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \end{bmatrix}  
$$  
Solving: $x = y$, so an eigenvector is **$v_1 = \begin{bmatrix} 1 \\ 1 \end{bmatrix}$**.  

For $\lambda_2 = 1$:  
$$  
\begin{bmatrix} 1 & 1 \\ 1 & 1 \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \end{bmatrix}  
$$  
Solving: $x = -y$, so an eigenvector is **$v_2 = \begin{bmatrix} 1 \\ -1 \end{bmatrix}$**.  

✅ **Eigenpairs:**  
$$  
\lambda_1 = 3, \quad v_1 = \begin{bmatrix} 1 \\ 1 \end{bmatrix}  
$$  
$$  
\lambda_2 = 1, \quad v_2 = \begin{bmatrix} 1 \\ -1 \end{bmatrix}  
$$  

📌 **Interpretation:**  
- **$v_1 = [1,1]$** is **stretched by 3x**.  
- **$v_2 = [1,-1]$** is **stretched by 1x (remains the same)**.  

## **📌 What is a Linear Transformation?**  

✅ A **linear transformation** is a function that **maps vectors to new vectors**, preserving **addition** and **scalar multiplication**.  
✅ It can represent **scaling, rotation, shearing, and reflection**.  

📌 **Matrix Representation:** If a transformation $T(x)$ is **linear**, there exists a matrix $A$ such that:  

$$  
T(x) = Ax  
$$  

✅ **Examples of Linear Transformations & Their Meaning:**  

| Transformation                  | Matrix $A$                                        | Meaning                         |  
|---------------------------------|---------------------------------------------------|---------------------------------|  
| **Scaling**                     | $\begin{bmatrix} k & 0 \\ 0 & k \end{bmatrix}$   | Stretches space by $k$          |  
| **Rotation (90° CCW)**          | $\begin{bmatrix} 0 & -1 \\ 1 & 0 \end{bmatrix}$  | Rotates counterclockwise        |  
| **Reflection over x-axis**      | $\begin{bmatrix} 1 & 0 \\ 0 & -1 \end{bmatrix}$  | Flips over x-axis               |  
| **Shear (x-direction)**         | $\begin{bmatrix} 1 & k \\ 0 & 1 \end{bmatrix}$   | Tilts space                    |  

✅ **Eigenvalues in Linear Transformations:**  
- **Scaling transformation:** Eigenvalues represent **how much the space is stretched**.  
- **Rotation transformation:** Eigenvalues are **complex numbers** (rotation does not have real eigenvalues).  
- **Reflection transformation:** Eigenvalues are **+1 (same direction) and -1 (flipped direction)**.  

## **🛠️ Python Code: Finding Eigenvalues & Eigenvectors**  

```python  

In [1]:
import numpy as np

# Define matrix
A = np.array([[2, 1], [1, 2]])

# Compute eigenvalues and eigenvectors
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvalues:", eigenvalues)
print("Eigenvectors:\n", eigenvectors)

Eigenvalues: [3. 1.]
Eigenvectors:
 [[ 0.70710678 -0.70710678]
 [ 0.70710678  0.70710678]]


```  

✅ **Output:**  
```  
Eigenvalues: [3. 1.]
Eigenvectors:
 [[ 0.707  0.707]
 [ 0.707 -0.707]]  
```  

📌 **Interpretation:** The eigenvectors are **normalized** (scaled to length 1).  

## **🚀 Real-World Applications of Eigenvalues & Eigenvectors**  

✅ **Machine Learning (PCA - Principal Component Analysis):** Reducing data dimensions 📊  
✅ **Computer Vision (Image Compression):** Eigenvectors define important features in images 📷  
✅ **Physics (Quantum Mechanics):** Finding energy levels of particles 🔬  
✅ **Vibrations & Stability Analysis:** Engineering applications for stability analysis ⚙️  
✅ **Google PageRank Algorithm:** Uses eigenvectors to rank web pages 🌍  

## **🔥 Summary**  

1️⃣ **Eigenvectors stay in the same direction during a transformation, eigenvalues are the scaling factors.**  
2️⃣ **Formula:** $Av = \lambda v$, where $A$ is a matrix, $v$ is an eigenvector, and $\lambda$ is an eigenvalue.  
3️⃣ **Linear transformations (scaling, rotation, reflection) determine how space is changed.**  
4️⃣ **Used in PCA, computer vision, quantum mechanics, and engineering stability analysis.**  